SYCL/Reduction/reduction_nd_N_queue_shortcut.cpp

```diff
@@ -0,0 +1,144 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+
+// `Group algorithms are not supported on host device.` on NVidia.
+// XFAIL: hip_nvidia
+
+// RUNx: %HOST_RUN_PLACEHOLDER %t.out
+// TODO: Enable the test for HOST when it supports ext::oneapi::reduce() and
+// barrier()
+
+// This test only checks that the method queue::parallel_for() accepting
+// reduction, can be properly translated into queue::submit + parallel_for().
+
+#include "reduction_utils.hpp"
+
+using namespace sycl;
+
+template <typename T, int I> class KName;
+
+enum TestCase { NoDependencies, Dependency, DependenciesVector };
+
+template <typename T> T *allocUSM(queue &Q, size_t Size) {
+  if (!Q.get_device().has(getUSMAspect(usm::alloc::shared)))
+    return nullptr;
+
+  return malloc_shared<T>(Size, Q);
+}
+
+template <typename RangeT>
+void printNVarsTestLabel(bool IsSYCL2020, const RangeT &Range,
+                         bool ToCERR = false) {
+  std::ostream &OS = ToCERR ? std::cerr : std::cout;
+  std::string Mode = IsSYCL2020 ? "SYCL2020" : "ONEAPI  ";
+  OS << (ToCERR ? "Error" : "Start") << ": Mode=" << Mode
+     << ", Range=" << Range;
+  if (!ToCERR)
+    OS << std::endl;
+}
+
+template <typename T1, typename T2, TestCase TC, int Dims, typename BOpT1,
+          typename BOpT2>
+int test(queue &Q, BOpT1 BOp1, BOpT2 BOp2, const nd_range<Dims> &Range) {
+  printNVarsTestLabel(true /*SYCL2020*/, Range);
+
+  size_t NElems = Range.get_global_range().size();
+  T1 *Sum1 = allocUSM<T1>(Q, 1);
+  T2 *Sum2 = allocUSM<T2>(Q, 1);
+  T1 *Arr1 = allocUSM<T1>(Q, NElems);
+  T2 *Arr2 = allocUSM<T2>(Q, NElems);
+  if (!Sum1 || !Sum2 || !Arr1 || !Arr2) {
+    sycl::free(Sum1, Q);
```

Comment:
Is it safe to free these if they are `nullptr`?

Reply:
free(nullptr) has no effect

Reply:
Perfect, thanks!